### DISCUSSION TITLE
WINDOW関数の具体的な使用例

### DISCUSSION THREAD

**Discussion by yuyakawai0411** [kakeru-one/web-engineering-roundtable#88] [General]
## 1.シチュエーション
[こちら](https://github.com/kakeru-one/web-engineering-roundtable/discussions/87)と同じシチュエーションに追加で、company_settingsにある会社名もついでに表示して欲しいと言われたケース
以下のSQLのままだと、company_settings.nameを取得することが出来ない

```sql
SELECT C.company_master_id
FROM tenant_payloads AS TP
JOIN leasings AS L ON L.id = TP.leasing_id
JOIN company_settings AS C ON C.company_master_id = L.company_master_id
GROUP BY C.company_master_id
HAVING sum(
  CASE 
    WHEN TP.send_type = 0 THEN 1
    WHEN TP.send_type = 1 THEN 0
  END
) = 0;
```

## 2.集合思考で考える
以下のようにマイページ招待をしたことがある会社をcompany_maseter_id毎にパーティションで区切れば対象の会社が抽出できそう

[![Image from Gyazo](https://i.gyazo.com/aa820415e83ddc12e175e88c708c4430.png)](https://gyazo.com/aa820415e83ddc12e175e88c708c4430)

## 3.SQLクエリで表現する
集合思考で考えた結果をSQLクエリで表現する。区切りはWINDOW関数のPARTITION BYで表現できる。

```sql
SELECT
  TP.id,
  C.company_master_id,
  C.name,
  MIN(TP.send_type) OVER (PARTITION BY C.company_master_id) AS 'min_send_type'
FROM tenant_payloads AS TP
JOIN leasings AS L ON L.id = TP.leasing_id
JOIN company_settings AS C ON C.company_master_id = L.company_master_id
```

| tenant_payload.id | company_master_id | name | min_send_type |
| --- | --- | --- | --- | 
| 1 | 1 | 河合会社 | 0 |
| 2 | 1 | 河合会社 | 0 |
| 3 | 1 | 河合会社 | 0 |
| 4 | 2 | 越智会社 | 1 |
| 5 | 2 | 越智会社 | 1 |
| 6 | 2 | 越智会社 | 1 |

## 4.send_typeの最小値が1の会社を抽出する

```sql
SELECT DISTINCT(company_master_id), name
FROM (
  SELECT
    C.company_master_id,
    C.name,
    MIN(TP.send_type) OVER (PARTITION BY C.company_master_id) AS 'min_send_type'
  FROM tenant_payloads AS TP
  JOIN leasings AS L ON L.id = TP.leasing_id
  JOIN company_settings AS C ON C.company_master_id = L.company_master_id
) AS subquery
WHERE min_send_type = 1;
```

## コラム
もし会社とsent_type毎に何件送信したかの数をカウントしたいとしたら

### 頭の中の集合思考
[![Image from Gyazo](https://i.gyazo.com/b90513d650bf3568dcb4f0b62831e826.png)](https://gyazo.com/b90513d650bf3568dcb4f0b62831e826)

### SQLに起こす

RANKではなくROW_NUMBERを使っているところがポイント。RANKだと同じ値のとき、同じランク付けをしてしまうため。

```sql
SELECT name, send_type, MAX(count_by_send_type)
FROM(
  SELECT
    C.company_master_id,
    C.name,
    TP.send_type,
    ROW_NUMBER() OVER (PARTITION BY C.company_master_id, TP.send_type) AS 'count_by_send_type'
  FROM tenant_payloads AS TP
  JOIN leasings AS L ON L.id = TP.leasing_id
  JOIN company_settings AS C ON C.company_master_id = L.company_master_id
) AS subquery
GROUP BY name, send_type;
```

### サブクエリの結果
| company_master_id | name | send_type | count_by_send_type |
| --- | --- | --- | --- |
| 1 | 河合会社 | 0 | 1 |
| 1 | 河合会社 | 0 | 2 |
| 1 | 河合会社 | 0 | 3 |
| 2 | 越智会社 | 0 | 1 |
| 2 | 越智会社 | 0 | 2 |
| 2 | 越智会社 | 1 | 1 |
| 2 | 越智会社 | 1 | 2 |
| 2 | 越智会社 | 1 | 3 |

**Comment by kakeru-one** [comment]
```sql
SELECT
  DISTINCT(C.company_master_id),
  C.name,
  MIN(TP.send_type) OVER (PARTITION BY C.company_master_id) AS 'min_send_type'
FROM tenant_payloads AS TP
JOIN leasings AS L ON L.id = TP.leasing_id
JOIN company_settings AS C ON C.company_master_id = L.company_master_id
WHERE min_send_type = 1;
```

**Reply by kakeru-one** [reply]
select句は、where句の後に実行するので、min_send_typeは条件に含められない。

**Comment by sho-work** [comment]
https://dev.mysql.com/doc/refman/8.0/ja/window-function-descriptions.html#function_rank
